From 99c65ab84080d35c1ba1df6efd5641ad3ab9e1e8 Mon Sep 17 00:00:00 2001
From: Vladimir Dvorak <vladimir.dvorak@jetbrains.com>
Date: Sat, 20 Feb 2021 15:47:47 +0100
Subject: [PATCH 26/45] Add ClassLoaderDataGraph_lock to define new class in
 enhanced redefiniton

ClassLoaderDataGraph locking for introduced in redefinition in
java.version>11
---
 src/hotspot/share/classfile/systemDictionary.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/hotspot/share/classfile/systemDictionary.cpp b/src/hotspot/share/classfile/systemDictionary.cpp
index 35e2f376870..c628cb9da83 100644
--- a/src/hotspot/share/classfile/systemDictionary.cpp
+++ b/src/hotspot/share/classfile/systemDictionary.cpp
@@ -1443,7 +1443,9 @@ void SystemDictionary::define_instance_class(InstanceKlass* k, InstanceKlass* ol
   if (is_redefining) {
     // Update all dictionaries containing old_class to new_class
     // outcome must be same as result of standard redefinition, that does not create a new Klass
+    ClassLoaderDataGraph_lock->lock();
     bool ok = ClassLoaderDataGraph::dictionary_classes_do_update_klass(name_h, k, old_klass);
+    ClassLoaderDataGraph_lock->unlock();
     assert (ok, "must have found old class and updated!");
   }
   check_constraints(name_hash, k, class_loader, !is_redefining, CHECK);
-- 
2.23.0

